rfd/0137-unified-resource-view.md

```diff
@@ -0,0 +1,121 @@
+```go
+func (u *uiResourceCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := u.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = u.getAndUpdateDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	//...
+
+	u.defineCollectorAsInitialized()
+	return nil
+}
+```
+
+If the collector is uninitialized or stale, we will resort to the upstream and send Nodes only.
+
+### ListUnifiedResources grpc endppint
+We will add a new grpc endpoint, `ListUnifiedResources`. It will function pretty closes to how `ListResources` works, except without needing to differentiate between `RequestType`, as every request is a unified resource request. Originally, the thought of just extending `ListResources`, but because of the many new filters that will be added to the request, we don't want to muddy up the `ListResourcesRequest`. Some of the new filters include which types to include (can be multiple), subtypes, etc etc. 
+
+This endpoint will pull resources from our unified resource cache and run through the same filtering/rbac as `ListResources`
+
+### Backward Compatibility
+
+We will leave `ListResources` alone and keep the old individual resource endpoints available until we've vetted the unified resource view. We will remove the individual navigation items from the web ui, but leave the routes/api endpoints until we've gained confidence in the new flow. 
+
+### UX
+The UI for the web will be changing drastically given the fact that all resources will be on the same page and we will be using a Card layout rather than the current Table. However, besides the visual changes, the actual resources listed will still be interacted with the same way. You can still click "Connect" on servers and databases, or Login for apps. The individual functionality for each "Item" will remain the same. Labels will be truncated but viewable with an expand button.
+
+
+
+https://github.com/gravitational/teleport/assets/5201977/9c859a2a-784d-4805-9ba7-9e37a8a4ce77
+
+
+
+Also, filters will be added. Some filters will just be a "pseudo-filter" in the sense that it'll just help populate an advanced search string, (example, clicking a label), but a few new actual filters will be present in the requests. The biggest of them is the `Kinds` array, which will contain a list of kinds that you want to be returned. ["node","app"] will only return nodes and apps. If left empty (the default), all resource kinds will be returned
+
+![Screenshot 2023-07-25 at 4 18 19 PM](https://github.com/gravitational/teleport/assets/5201977/0a58e6bc-b94f-44aa-9a63-56d4991a395c)
+
+
+This page will also use "infinite scrolling" rather than pagination. We will employ the use of a loading skeleton as we wait for the data to come in and then at a certain scroll distance from the bottom (not the exact bottom, but maybe 200 or so pixels up) we will fetch more. [example of skeleton component](https://codepen.io/JCLee/pen/dyPejGV)
```

Comment:
i never worked with infinite scrolling, but will we be off loading the previously fetched resources?

Reply:
We will not be offloading previously fetch resources. Just appending new ones at the bottom of the list